### PR TITLE
Verbose Error Handling

### PR DESCRIPTION
**Pull request by ferrell-code**
## Description

- Deletes suspend manta-pay. This is because the suspender in its current form cannot work due to the transactional nature of extrinsics. If an extrinsic fails all storage changes are rolled back including pausing manta-pay. 
- Adds lots of errors to be more specific
- Moves error handling logic to its own file to be more tidy

---

Before we can approve this PR for merge, please make sure that **all** the following items have been checked off:
- [x] Connected to an issue with discussion and accepted design using zenhub "Connect issue" button below
- [x] Added **one** label out of the `L-` group to this PR
- [x] Added **one or more** labels from the `A-` and `C-` groups to this PR
- [x] Explicitly labelled `A-calamari`, `A-dolphin` and/or `A-manta` if your changes are meant for/impact either of these (CI depends on it)
- [x] Re-reviewed `Files changed` in the Github PR explorer.

Situational Notes:
- If adding functionality, write unit tests!
- If importing a new pallet, choose a proper module index for it, and allow it in `BaseFilter`. Ensure **every** extrinsic works from front-end. If there's corresponding tool, ensure both work for each other.
- If needed, update our Javascript/Typescript APIs. These APIs are officially used by exchanges or community developers.
- If modifying existing runtime storage items, make sure to implement storage migrations for the runtime and test them with `try-runtime`. This includes migrations inherited from upstream changes, and you can search the diffs for modifications of `#[pallet::storage]` items to check for any.

### CODE REVIEW

File: pallets/manta-pay/src/errors.rs [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+impl<T> From<ReceiverPostError<ReceiverLedgerError<T>>> for Error<T>
+where
+    T: Config,
+{
+    #[inline]
+    fn from(err: ReceiverPostError<ReceiverLedgerError<T>>) -> Self {
+        match err {
+            ReceiverPostError::AssetRegistered => Self::AssetRegistered,
+            ReceiverPostError::UnexpectedError(e) => match e {
+                ReceiverLedgerError::<T>::AssetRegistered => Self::AssetRegistered,
```
Comment (ferrell-code) [inline pallets/manta-pay/src/errors.rs:119]: Exhaustive error handling creates duplicates... It functionally is fine, but could change `manta-rs` to get rid of this

File: pallets/manta-pay/src/errors.rs [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+impl<T> From<SenderPostError<SenderLedgerError>> for Error<T>
+where
+    T: Config,
+{
+    #[inline]
+    fn from(err: SenderPostError<SenderLedgerError>) -> Self {
+        match err {
+            SenderPostError::AssetSpent => Self::AssetSpent,
+            SenderPostError::InvalidUtxoAccumulatorOutput => Self::InvalidUtxoAccumulatorOutput,
```
Comment (ferrell-code) [inline pallets/manta-pay/src/errors.rs:186]: same here

File: pallets/manta-pay/src/lib.rs [hunk 3]
```diff
@@ -953,59 +789,6 @@ where
...
-    /// Asset Registered Error
-    ///
-    /// The asset has already been registered with the ledger.
-    AssetRegistered,
-
-    /// Type Marker Parameter
-    Marker(PhantomData<T>),
-}
-
 impl<T> From<ReceiverLedgerError<T>> for ReceiverPostError<ReceiverLedgerError<T>>
```
Comment (zqhxuyuan) [inline pallets/manta-pay/src/lib.rs]: move to errors.rs

File: runtime/dolphin/src/assets_config.rs [hunk 4]
```diff
@@ -184,5 +184,4 @@ impl pallet_manta_pay::Config for Runtime {
     type WeightInfo = weights::pallet_manta_pay::SubstrateWeight<Runtime>;
     type AssetConfig = DolphinAssetConfig;
     type PalletId = MantaPayPalletId;
-    type Suspender = ();
```
Comment (ghzlatarev) [inline runtime/dolphin/src/assets_config.rs:187]: can you also add why we're deleting the suspender in the description, please.

File: pallets/manta-pay/src/errors.rs [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    T: Config,
+{
+    #[inline]
+    fn from(value: TransferLedgerError<T>) -> Self {
+        match value {
+            TransferLedgerError::InvalidProof => Self::InvalidProof,
+            TransferLedgerError::InvalidTransferShape => Self::InvalidShape,
+            TransferLedgerError::SenderLedgerError(err) => Self::Sender(err.into()),
+            TransferLedgerError::ReceiverLedgerError(err) => Self::Receiver(err.into()),
+            err => Self::UnexpectedError(err),
```
Comment (ghzlatarev) [inline pallets/manta-pay/src/errors.rs:360]: I thought we wanted to get rid of or at least minimize the unexpected/internal-ledger-error
This catch all still won't describe/trace the FungibleLedgerError -> TransferLedgerError -> TransferPostError -> PalletError transform.
Maybe we should also add a debug log ?

Comment (ferrell-code) [reply pallets/manta-pay/src/errors.rs:360]: Actually this `UnexpectedError` is from manta-rs there isn't really a way to remove it without altering manta-rs. 

It gets resolved in an exhaustive pattern to the pallet error type here.
```rust 
impl<T> From<TransferPostError<T>> for Error<T>
```

I agree it is a bit convoluted but functionally it should point us to the exact error. This is the same issue as my two comments above

**Review by ghzlatarev — DISMISSED** [review]

**Review by ghzlatarev — DISMISSED** [review]

**Review by ghzlatarev — APPROVED** [review]

**Review by zqhxuyuan — APPROVED** [review]

**Review by bhgomes — APPROVED** [review]

**Pull request merged**